[bl0940] document new configuration options (#4744)
OttoWinter/esphomedocs: components/sensor/bl0940.rst

    this->energy_reference_ = this->power_reference_ * 3600000 / (1638.4 * 256);

If ``legacy_mode`` is enabled the reference values will be set according to initial implementation.

.. code-block:: cpp

    this->voltage_reference_ = 33000;
    this->current_reference_ = 275000;
    this->power_reference_   = 1430;
    this->energy_reference_  = 3.6e6 / 297;

Any of the reference values may be manually configured; user-specified values will take precedence over the calculated/predefined values.

Theoretically, configuring ``voltage_reference`` and ``current_reference`` should be enough to calibrate the sensors, but, for completeness, ``power_reference`` and ``energy_reference`` are also configurable.

.. code-block:: yaml

    # custom references
    voltage_reference: 17158.9201389365
    current_reference: 266013.136288998
    power_reference:   713.104696500825
    energy_reference:  6120.6267056536

Online Calibration
------------------

It is possible to adjust the sensor readings without need to update sensor configuration using the provided number platform.

.. code-block:: yaml

>     # Example configuration entry
>     sensor:
>       - platform: bl0940
>         id: bl0940_id
>         uart_id: uart_bus
>         ...
> 

Reviewer: ```suggestion
```

Author: removed

>     number:

Reviewer: Ideally each entity type goes on its own page in the relevant folder. eg: `/components/number/bl0940.rst`

Author: created sections for each platform to keep everything together

- **max_value** (*Optional*, float): The maximum value this number can be. Maiximal allowed value is ``50`` defaults to ``10``.
- **step** (*Optional*, float): The granularity with which the number can be set. Defaults to ``0.1``.
- **restore_value** (*Optional*, boolean): Saves and loads the state to RTC/Flash. Defaults to ``True``
- All other options from :ref:`Number <config-number>`.

.. note::

    There are some fixed values defined:

    - **unit_of_measurement**: `UNIT_PERCENT`
    - **mode**: `BOX`
    - **initial_value**: ``0``

Calibration details:
********************
Calibration factor is calculated in ``%`` based on submitted value from the configured ``calibration`` component.

.. code-block:: cpp

    float BL0940::calculate_calibration_value_(float state) { return (100 + state) / 100; }

When changing calibration values on the UI all calibrated reference values will be recalculated.

``voltage_calibration`` and ``current_calibration`` will affect ``power_reference`` and ``energy_reference`` if they have not been specified manually.

Reset Calibration:

> ******************

Reviewer: https://developers.esphome.io/contributing/docs/#headers

Author: updated headers to create proper structure

To reset all external calibration values to their initial state, you can use the calibration reset button:

.. code-block:: yaml

    button:
      - platform: bl0940
        bl0940_id: bl0940_id
        name: Reset Calibration

Configuration variables:
************************
- **bl0940_id** (*Required*, :ref:`config-id`): Manually specify the ID of the `BL0940 Sensor Component`
- All other options from :ref:`Button <config-button>`.

.. figure:: images/bl0940-cal.png
    :align: center
    :width: 80.0%

> Sample full configuration:

Reviewer: I personally dont think this is required on this page.

Author: removed

--------------------------

.. code-block:: yaml

    uart:
      id: uart_bus

>       tx_pin: TX
>       rx_pin: RX

Reviewer: We don't add specific pins or alises
```suggestion
      tx_pin: GPIOXX
      rx_pin: GPIOXX
```

Author: changed on line 24-25 as this section is now gone

By default all the reference values will be calculated using the above values in the exact order shown below.

.. code-block:: cpp

    this->voltage_reference_ = 79931 / this->vref_ * (this->r_one_ * 1000) / (this->r_one_ + this->r_two_);
    this->current_reference_ = 324004 * this->r_shunt_ / this->vref_;

    // if voltage reference and current_reference have been configured
    this->power_reference_ = this->voltage_reference_ * this->current_reference_ * 4046 / 324004 / 79931;
    // if not, we use this formula based on provided formula in the specification
    this->power_reference_ = 4046 * this->r_shunt_ * this->r_one_ * 1000 / this->vref_ / this->vref_ / (this->r_one_ + this->r_two_);

    this->energy_reference_ = this->power_reference_ * 3600000 / (1638.4 * 256);

If ``legacy_mode`` is enabled the reference values will be set according to initial implementation.

.. code-block:: cpp

    this->voltage_reference_ = 33000;
    this->current_reference_ = 275000;
    this->power_reference_   = 1430;
    this->energy_reference_  = 3.6e6 / 297;

> Any of the reference values is configurable and will take precedence over the calculated or predefined values.
> 
> Theoreticaly configuring ``voltage_reference`` and ``current_reference`` should be enough to calibrate the sensors but for completeness ``power_reference`` and ``energy_reference`` are also configurable.

Reviewer 2: ```suggestion
Any of the reference values may be manually configured; user-specified values will take precedence over the calculated/predefined values.

Theoretically, configuring ``voltage_reference`` and ``current_reference`` should be enough to calibrate the sensors, but, for completeness, ``power_reference`` and ``energy_reference`` are also configurable.
```

- **max_value** (*Optional*, float): The maximum value this number can be. Maiximal allowed value is ``50`` defaults to ``10``.
- **step** (*Optional*, float): The granularity with which the number can be set. Defaults to ``0.1``.
- **restore_value** (*Optional*, boolean): Saves and loads the state to RTC/Flash. Defaults to ``True``
- All other options from :ref:`Number <config-number>`.

.. note::

    There are some fixed values defined:

    - **unit_of_measurement**: `UNIT_PERCENT`
    - **mode**: `BOX`
    - **initial_value**: ``0``

Calibration details:
********************
Calibration factor is calculated in ``%`` based on submitted value from the configured ``calibration`` component.

.. code-block:: cpp

    float BL0940::calculate_calibration_value_(float state) { return (100 + state) / 100; }

When changing calibration values on the UI all calibrated reference values will be recalculated.

``voltage_calibration`` and ``current_calibration`` will affect ``power_reference`` and ``energy_reference`` if they have not been specified manually.

Reset Calibration:
******************

> To reset all external calibration values to their initial state you can use the provided calibration reset button
> 
> .. code-block:: yaml
> 
>     # Example configuration entry
>     sensor:
>       - platform: bl0940
>         id: bl0940_id
>         uart_id: uart_bus
>         ...
> 
>     button:
>       - platform: bl0940
>         bl0940_id: bl0940_id
>         name: Reset Calibration

Reviewer 2: ```suggestion
To reset all external calibration values to their initial state, you can use the calibration reset button:

.. code-block:: yaml

    button:
      - platform: bl0940
        bl0940_id: bl0940_id
        name: Reset Calibration
```

By default all the reference values will be calculated using the above values in the exact order shown below.

.. code-block:: cpp

    this->voltage_reference_ = 79931 / this->vref_ * (this->r_one_ * 1000) / (this->r_one_ + this->r_two_);
    this->current_reference_ = 324004 * this->r_shunt_ / this->vref_;
    
    // if voltage reference and current_reference have been configured
    this->power_reference_ = this->voltage_reference_ * this->current_reference_ * 4046 / 324004 / 79931;
    // if not, we use this formula based on provided formula in the specification
    this->power_reference_ = 4046 * this->r_shunt_ * this->r_one_ * 1000 / this->vref_ / this->vref_ / (this->r_one_ + this->r_two_);
    
    this->energy_reference_ = this->power_reference_ * 3600000 / (1638.4 * 256);

If ``tuya_mode`` is enabled the reference values will be set according to initial implementation.

.. code-block:: cpp

    this->voltage_reference_ = 33000;
    this->current_reference_ = 275000;
    this->power_reference_   = 1430;
    this->energy_reference_  = 3.6e6 / 297;

> Any of the reference values is configurable and will take presedence over the calculated or predefined values.

Reviewer 3: precedence

Author: fixed typo
